### PR TITLE
CB-4: Comments

### PR DESCRIPTION
**Pull request by Narzew**

### CODE REVIEW

File: app/controllers/api/v1/comments_controller.rb [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+      def show
+        render json: CommentSerializer.new(@comment).to_h
+      end
+
+      def create
+        @comment = @recipe.comments.build(comment_params)
+        if @comment.save
+          render json: @comment
+        else
+          render json: { errors: @comment.errors.to_s }, status: :unprocessable_entity
```
Comment (PatrykTyszka) [inline app/controllers/api/v1/comments_controller.rb:21]: I don't get this? As I know @comment.errors.to_s returns sth like this: `"#<ActiveModel::Errors:0x00007fc621adde20>"` so maybe would be better to send @comment.errors.full_messages ?

File: app/serializers/recipe_serializer.rb [hunk 2]
```diff
@@ -26,4 +26,12 @@ def price
 
     object.price
   end
+
+  def comments
+    if @instance_options[:context][:action] == 'show'
```
Comment (PatrykTyszka) [inline app/serializers/recipe_serializer.rb:31]: so if you want to use this Serializer without context I would go with this solution ->
```
@instance_options.fetch(:context, {})[:action] == 'show'
```
You won't get an exception if you forget to put context inside.

Besides I'm not sure if this solution is OK. Maybe, for now, we can leave it as is coz we have only one difference ( display comments only for SHOW action ) but if you'll have more than one check I would go with a separate Serializer to make this class more readable. You need to remember that DRY is important but READABILITY is better for sure ;)

File: config/routes.rb [hunk 3]
```diff
@@ -2,7 +2,9 @@
   # For details on the DSL available within this file, see https://guides.rubyonrails.org/routing.html
   namespace :api do
     namespace :v1 do
-      resources :recipes
+      resources :recipes do
+        resources :comments
```
Comment (PatrykTyszka) [inline config/routes.rb:6]: As I know resources method creates all routes including Edit and New. We need to create only these routes which we're using so please add `only` and show up exactly what you want to create.

File: spec/controllers/api/v1/comments_controller_spec.rb [hunk 4]
```diff
@@ -0,0 +1,4 @@
+require 'rails_helper'
+
+describe Api::V1::CommentsController do
+end
```
Comment (PatrykTyszka) [inline spec/controllers/api/v1/comments_controller_spec.rb:4]: missing specs ;)

File: spec/factories/comments.rb [hunk 5]
```diff
@@ -0,0 +1,7 @@
+FactoryBot.define do
+  factory :comment do
+    author { Faker::Name.name }
+    body { Faker::Hacker.say_something_smart }
+    recipe_id { rand(1..4) }
```
Comment (PatrykTyszka) [inline spec/factories/comments.rb:5]: wrong way! :D I agree that we can use rand for some number like `level` or sth but for recipe_id ( relation ) I don't agree for sure. Please create this relation in a proper way.

File: app/controllers/api/v1/recipes_controller.rb [hunk 6]
```diff
@@ -3,38 +3,44 @@ module V1
...
         else
-          render json: { errors: @recipe.errors.messages }, status: :unprocessable_entity
+          render json: { errors: @recipe.errors.full_messages }, status: :unprocessable_entity
         end
       end
 
+      def comments
+        render json: { data: ActiveModel::SerializableResource.new(Comment.where(recipe_id: params[:recipe_id]),
+                                                                   each_serialize: CommentSerializer) }
+      end
```
Comment (h4b00) [inline app/controllers/api/v1/recipes_controller.rb]: ?

File: app/models/recipe.rb [hunk 7]
```diff
@@ -9,6 +23,8 @@ class Recipe < ApplicationRecord
 
   # == Relationships ========================================================
 
+  has_many :comments
```
Comment (h4b00) [inline app/models/recipe.rb:26]: `dependent: :destroy`
https://guides.rubyonrails.org/association_basics.html#why-associations-questionmark

File: app/serializers/recipe_serializer.rb [hunk 8]
```diff
@@ -26,4 +40,14 @@ def price
 
     object.price
   end
+
+  def comments
+    return nil if @instance_options.nil? or @instance_options[:context].nil? or @instance_options[:context][:action].nil?
```
Comment (h4b00) [inline app/serializers/recipe_serializer.rb]: 1.  do not use `or` use `||`
2. `context[:action]` is not so much descriptive, use `context[:include_comments]`
3. What is the difference here?
```
      return nil if object.comments.empty?
      return nil unless object.comments
```

File: spec/models/comment_spec.rb [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+#
+#  index_comments_on_recipe_id  (recipe_id)
+#
+require 'rails_helper'
+
+RSpec.describe Comment, type: :model do
+
+  #recipe = create :recipe
+  #comment = create :comment, recipe: recipe
+
```
Comment (h4b00) [inline spec/models/comment_spec.rb]: Unnecessary

File: db/migrate/20210804092639_create_comments.rb [hunk 10]
```diff
@@ -0,0 +1,11 @@
+class CreateComments < ActiveRecord::Migration[6.1]
+  def change
+    create_table :comments do |t|
+      t.string :author
+      t.string :body
```
Comment (h4b00) [inline db/migrate/20210804092639_create_comments.rb:5]: I would protect it before being null

File: spec/models/comment_spec.rb [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+#
+# Indexes
+#
+#  index_comments_on_recipe_id  (recipe_id)
+#
+require 'rails_helper'
+
+RSpec.describe Comment, type: :model do
+  # recipe = create :recipe
+  # comment = create :comment, recipe: recipe
```
Comment (h4b00) [inline spec/models/comment_spec.rb:20]: ?

File: spec/controllers/api/v1/recipes_controller_spec.rb [hunk 12]
```diff
@@ -5,6 +5,7 @@
   describe '[GET] #index' do
     before do
       create_list(:recipe, 5)
+      # byebug
```
Comment (h4b00) [inline spec/controllers/api/v1/recipes_controller_spec.rb:8]: ?

File: spec/factories/comments.rb [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+#
+# Indexes
+#
+#  index_comments_on_recipe_id  (recipe_id)
+#
+FactoryBot.define do
+  factory :comment do
+    author { Faker::Name.name }
+    body { Faker::Hacker.say_something_smart }
+    recipe_id { recipe }
```
Comment (h4b00) [inline spec/factories/comments.rb:20]: recipe
```suggestion
    body { Faker::Hacker.say_something_smart }
    recipe
```

File: app/serializers/recipe_serializer.rb [hunk 14]
```diff
@@ -26,4 +40,15 @@ def price
...
   end
+
+  def comments
+    return nil if @instance_options.nil? || @instance_options[:context].nil? || @instance_options[:context][:action].nil?
+
+    if @instance_options.fetch(:context, {})[:action] == 'show'
+      return nil if object.comments.empty?
+      return nil unless object.comments
+
+      object.comments
```
Comment (h4b00) [inline app/serializers/recipe_serializer.rb:51]: return unless @instance_options
```suggestion
    return [] unless @instance_options[:include_comments]
    
    object.comments
```

File: app/controllers/api/v1/recipes_controller.rb [hunk 15]
```diff
@@ -3,35 +3,36 @@ module V1
...
       before_action :set_recipe, except: %i[create index]
       def index
-        render json: { data: ActiveModel::SerializableResource.new(Recipe.all, each_serialize: RecipeSerializer) }
+        render json: { data: ActiveModel::SerializableResource.new(Recipe.all, each_serialize: RecipeSerializer,
+                                                                               context: { action: 'index' }) }
       end
 
       def show
-        render json: RecipeSerializer.new(@recipe).to_h
+        render json: RecipeSerializer.new(@recipe, context: { action: 'show' }).to_h
```
Comment (h4b00) [inline app/controllers/api/v1/recipes_controller.rb:11]:
```suggestion
        render json: RecipeSerializer.new(@recipe, include_options: true).to_h
```

**Review by h4b00 — CHANGES_REQUESTED** [review]

File: app/controllers/api/v1/recipes_controller.rb (hunk 6 above)
Comment (Narzew) [reply app/controllers/api/v1/recipes_controller.rb]: Already deleted

File: app/controllers/api/v1/recipes_controller.rb [hunk 16]
```diff
@@ -3,35 +3,36 @@ module V1
...
       before_action :set_recipe, except: %i[create index]
       def index
-        render json: { data: ActiveModel::SerializableResource.new(Recipe.all, each_serialize: RecipeSerializer) }
+        render json: { data: ActiveModel::SerializableResource.new(Recipe.all, each_serialize: RecipeSerializer,
+                                                                               context: { include_comments: 'no' }) }
       end
 
       def show
-        render json: RecipeSerializer.new(@recipe).to_h
+        render json: RecipeSerializer.new(@recipe, context: { include_comments: 'show' }).to_h
```
Comment (h4b00) [inline app/controllers/api/v1/recipes_controller.rb]:
```suggestion
        render json: RecipeSerializer.new(@recipe, include_comments: true).to_h
```

File: app/serializers/comment_serializer.rb [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+
+  attributes :author, :body
+
+  def author
+    return UNKNOWN_VALUE unless object.author
+
+    object.author
+  end
+
+  delegate :body, to: :object
```
Comment (h4b00) [inline app/serializers/comment_serializer.rb:27]:
```suggestion
```

File: app/serializers/recipe_serializer.rb [hunk 18]
```diff
@@ -26,4 +40,17 @@ def price
...
+
+  def comments
+    if @instance_options.nil? || @instance_options[:context].nil? || @instance_options[:context][:include_comments].nil?
+      return nil
+    end
+
+    if @instance_options.fetch(:context, {})[:include_comments] == 'show'
+      return nil if object.comments.empty?
+      return nil unless object.comments
+
```
Comment (h4b00) [inline app/serializers/recipe_serializer.rb]:
```suggestion
 return [] unless @instance_options[:include_comments]
```

File: app/controllers/api/v1/recipes_controller.rb [hunk 19]
```diff
@@ -3,35 +3,36 @@ module V1
     class RecipesController < ApplicationController
       before_action :set_recipe, except: %i[create index]
       def index
-        render json: { data: ActiveModel::SerializableResource.new(Recipe.all, each_serialize: RecipeSerializer) }
+        render json: { data: ActiveModel::SerializableResource.new(Recipe.all, each_serialize: RecipeSerializer,
+                                                                               context: { include_comments: 'no' }) }
```
Comment (h4b00) [inline app/controllers/api/v1/recipes_controller.rb:7]:
```suggestion
                                                                               include_comments: false) }
```

**Review by h4b00 — CHANGES_REQUESTED** [review]

File: app/serializers/recipe_serializer.rb [hunk 20]
```diff
@@ -26,4 +40,17 @@ def price
...
+    if @instance_options.nil? || @instance_options[:context].nil? || @instance_options[:context][:include_comments].nil?
+      return nil
+    end
+
+    if @instance_options.fetch(:context, {})[:include_comments] == true
+      return nil if object.comments.empty?
+      return nil unless object.comments
+
+      object.comments
+    end
```
Comment (h4b00) [inline app/serializers/recipe_serializer.rb:54]:
```suggestion
    return [] unless @instance_options[:include_comments]

    object.comments
```

File: db/migrate/20210804092639_create_comments.rb [hunk 21]
```diff
@@ -0,0 +1,11 @@
+class CreateComments < ActiveRecord::Migration[6.1]
+  def change
+    create_table :comments do |t|
+      t.string :author
+      t.string :body, null: false
+      t.references :recipe
```
Comment (h4b00) [inline db/migrate/20210804092639_create_comments.rb:6]: null false?

File: app/models/comment.rb [hunk 22]
```diff
@@ -0,0 +1,43 @@
...
+#
+
+class Comment < ApplicationRecord
+  # == Constants ============================================================
+
+  # == Attributes ===========================================================
+
+  # == Extensions ===========================================================
+
+  include ActiveModel::Validations
```
Comment (h4b00) [inline app/models/comment.rb:24]: We don't need this

File: spec/controllers/api/v1/comments_controller_spec.rb [hunk 23]
```diff
@@ -0,0 +1,83 @@
...
+
+    before do
+      create_list(:comment, 5, recipe: recipe)
+      get :index, params: { recipe_id: recipe.id }
+    end
+
+    it 'returns all comments for a recipe' do
+      expect(JSON.parse(response.body)['data'].size).to eq(5)
+    end
+  end
```
Comment (h4b00) [inline spec/controllers/api/v1/comments_controller_spec.rb:15]: Not needed

File: app/controllers/api/v1/comments_controller.rb [hunk 24]
```diff
@@ -0,0 +1,56 @@
+module Api
+  module V1
+    class CommentsController < ApplicationController
+      before_action :set_recipe
+      before_action :set_comment, except: %i[index create]
+
+      def index
+        @comments = @recipe.comments.order(created_at: :desc)
+        render json: { data: ActiveModel::SerializableResource.new(@comments, each_serialize: CommentSerializer) }
+      end
```
Comment (h4b00) [inline app/controllers/api/v1/comments_controller.rb:10]: Not needed

File: app/controllers/api/v1/comments_controller.rb [hunk 25]
```diff
@@ -0,0 +1,56 @@
...
+
+      def show
+        render json: CommentSerializer.new(@comment).to_h
+      end
```
Comment (h4b00) [inline app/controllers/api/v1/comments_controller.rb:14]: Same as above

**Review by h4b00 — CHANGES_REQUESTED** [review]

File: app/controllers/api/v1/comments_controller.rb [hunk 26]
```diff
@@ -0,0 +1,47 @@
+module Api
+  module V1
+    class CommentsController < ApplicationController
+      before_action :set_recipe
+      before_action :set_comment, except: %i[index create]
+
+      def create
+        @comment = @recipe.comments.build(comment_params)
```
Comment (PatrykTyszka) [inline app/controllers/api/v1/comments_controller.rb:8]: why instance variable instead of local variable?

File: Gemfile [hunk 27]
```diff
@@ -48,12 +48,12 @@ end
 group :development do
   gem 'listen', '~> 3.3'
   # Spring speeds up development by keeping your application running in the background. Read more: https://github.com/rails/spring
+  gem 'annotate', '~> 3.1.1', require: false
```
Comment (PatrykTyszka) [inline Gemfile:51]: IMO this lib should be added in a separate PR.

Comment (mirecki) [reply Gemfile:51]: It was added by me and it was in separate PR or patch/ or approved by Adam to let be in one of my PR but I think first one. Dunno why it is displayed as his changes , probably due to late pull or smh

File: db/migrate/20210804092639_create_comments.rb (hunk 21 above)
Comment (Narzew) [reply db/migrate/20210804092639_create_comments.rb:6]: Used to protect field from being null

**Review by mirecki — APPROVED** [review]

File: db/migrate/20210804092639_create_comments.rb (hunk 21 above)
Comment (PatrykTyszka) [reply db/migrate/20210804092639_create_comments.rb:6]: @h4b00 If I good understood - @h4b00 wanted to know why you did not add null false here. am I right? :)

File: spec/controllers/api/v1/comments_controller_spec.rb [hunk 28]
```diff
@@ -0,0 +1,56 @@
...
+    end
+    let(:invalid_comment_params) do
+      {
+        recipe_id: recipe.id,
+        comment: {
+          author: '',
+          body: ''
+        }
+      }
+    end
```
Comment (PatrykTyszka) [inline spec/controllers/api/v1/comments_controller_spec.rb:23]: just put these let's to the proper context.

File: spec/controllers/api/v1/comments_controller_spec.rb [hunk 29]
```diff
@@ -0,0 +1,56 @@
...
+        end.not_to change(Comment, :count)
+      end
+    end
+  end
+
+  describe '[DELETE] #destroy' do
+    let!(:recipe) { create(:recipe) }
+    let!(:comment) { create(:comment, recipe: recipe) }
+    let(:expected_response) { CommentSerializer.new(comment).to_json }
+
```
Comment (PatrykTyszka) [inline spec/controllers/api/v1/comments_controller_spec.rb:46]: let(:call) { delete :destroy, params: { id: comment.id, recipe_id: recipe.id } }

and then...

expect { call }.to change ...

IMO this version is more readable.

File: spec/factories/comments.rb (hunk 13 above)
Comment (PatrykTyszka) [reply spec/factories/comments.rb:20]: does it work?

**Review by h4b00 — CHANGES_REQUESTED** [review]

File: config/routes.rb [hunk 30]
```diff
@@ -5,7 +5,9 @@
   # For details on the DSL available within this file, see https://guides.rubyonrails.org/routing.html
   namespace :api do
     namespace :v1 do
-      resources :recipes
+      resources :recipes do
+        resources :comments, only: %i[index show create destroy]
```
Comment (h4b00) [inline config/routes.rb:9]: where do you have index & show?

**Review by h4b00 — CHANGES_REQUESTED** [review]

**Review by h4b00 — APPROVED** [review]

**Review by PatrykTyszka — APPROVED** [review]

**Pull request merged**